### PR TITLE
Alias obsolete support

### PR DESCRIPTION
**Pull request by patriksvensson**
This is a WIP PR for obsoleting aliases.

### CODE REVIEW

**Comment by gep13** [comment]
@patriksvensson can you help me out here with some background?  What exactly is this, and why is it now being obseleted?  Don't think I have used this ability so far.

**Comment by patriksvensson** [comment]
@gep13 This is for if/when we need to depricate an alias in Cake. See issue #283 for more information.

**Comment by patriksvensson** [comment]
@gep13 Not only in Cake, but also Cake addins.

**Comment by gep13** [comment]
Ah, ok, now I see, thanks!

**Comment by gep13** [comment]
I am not really familiar with this area of the code yet, but this LGTM.  Will defer to @devlead for a more in depth review.

**Comment by patriksvensson** [comment]
@gep13 @devlead I appreciate that. Not quite finished yet so do not merge anything :)

**Comment by patriksvensson** [comment]
@gep13 @devlead This one is ready now. If someone of you review I'll can rebase to one commit and remove the WIP status (assuming no feedback) :smile:

**Comment by devlead** [comment]
@patriksvensson this looks solid, I will just do some testing with a add-in too before giving blessing :wink:

File: src/Cake.Core.Tests/Data/PropertyAliasGeneratorData.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        [CakePropertyAlias]
+        [Obsolete("Please use Foo.Bar instead.", true)]
+        public static int NonCached_Obsolete_Explicit(this ICakeContext context)
+        {
+            throw new NotImplementedException();
+        }
+
+
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Data/PropertyAliasGeneratorData.cs]: Double line breaks

File: src/Cake.Core.Tests/Unit/Scripting/CodeGen/PropertyAliasGeneratorTests.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+                // Then
+                Assert.Equal(expected, result);
+            }
+
+            [Theory]
+            [InlineData("Cached_Obsolete_ImplicitWarning_NoMessage")]
+            [InlineData("Cached_Obsolete_ImplicitWarning_WithMessage")]
+            [InlineData("Cached_Obsolete_ExplicitWarning_WithMessage")]
+            [InlineData("Cached_Obsolete_Explicit")]
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/Scripting/CodeGen/PropertyAliasGeneratorTests.cs]: Should be called `*_ExplicitError_WithMessage`

File: src/Cake.Core.Tests/Unit/Scripting/CodeGen/PropertyAliasGeneratorTests.cs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+                // Then
+                Assert.Equal(expected, result);
+            }
+
+            [Theory]
+            [InlineData("NonCached_Obsolete_ImplicitWarning_NoMessage")]
+            [InlineData("NonCached_Obsolete_ImplicitWarning_WithMessage")]
+            [InlineData("NonCached_Obsolete_ExplicitWarning_WithMessage")]
+            [InlineData("NonCached_Obsolete_Explicit")]
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/Scripting/CodeGen/PropertyAliasGeneratorTests.cs]: Should be called `*_ExplicitError_WithMessage`

File: src/Cake.Core.Tests/Unit/Scripting/CodeGen/MethodAliasGeneratorTests.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+                // Then
+                Assert.Equal(expected, result);
+            }
+
+            [Theory]
+            [InlineData("Obsolete_Implicit_Warning")]
+            [InlineData("Obsolete_Implicit_Warning_With_Message")]
+            [InlineData("Obsolete_Explicit_Warning")]
+            [InlineData("Obsolete_Explicit_Error")]
```
Comment (patriksvensson) [inline src/Cake.Core.Tests/Unit/Scripting/CodeGen/MethodAliasGeneratorTests.cs]: Should be called `*_ExplicitError_WithMessage`

**Comment by patriksvensson** [comment]
Ok, I think this one is ready now :smile:

**Pull request merged**